magellan/sampler/down_sample.py

```diff
@@ -0,0 +1,213 @@
+# coding=utf-8
+"""
+This module contains sampling related routines
+"""
+import os
+
+from random import randint
+import math
+import logging
+import random
+import pandas as pd
+import pyprind
+from magellan.utils.generic_helper import get_install_path
+import magellan.catalog.catalog_manager as cm
+
+logger = logging.getLogger(__name__)
+
+
+def _get_stop_words():
+    stop_words_set = set()
+    install_path = get_install_path()
+    dataset_path = os.sep.join([install_path, 'utils'])
+    stop_words_file = os.sep.join([dataset_path, 'stop_words.txt'])
+    with open(stop_words_file, "rb") as stopwords_file:
+        for stop_words in stopwords_file:
+            stop_words_set.add(stop_words.rstrip())
+
+    return stop_words_set
+
+
+# get string column list
+def _get_str_cols_list(table):
+    if len(table) == 0:
+        logger.error('_get_str_cols_list: Size of the input table is 0')
+        raise AssertionError('_get_str_cols_list: Size of the input table is 0')
+
+    cols = list(table.columns[table.dtypes == object])
+    col_list = []
+    for attr_x in cols:
+        col_list.append(table.columns.get_loc(attr_x))
+
+    return col_list
+
+
+# create inverted index from token to position
+def _inv_index(table):
+    """
+
+    This is inverted index function that builds inverted index of tokens on a table
+
+    """
+
+    # Get the stop words listed by user in the file specified by dataset path
+    stop_words = _get_stop_words()
+
+    # Extract indices of all string columns (if any) from the input dataframe
+    str_cols_ix = _get_str_cols_list(table)
+
+    inv_index = dict()
+    pos = 0
+
+    # For each row in the dataframe of the input table, we will fetch all string values from string column indices
+    # and will concatenate them. Next step would be to tokenize them using set and remove all the stop words
+    # from the list of tokens. Once we have the list of tokens, we will iterate through the list of tokens
+    # to identify its position and will create an inverted index.
+    for row in table.itertuples():
+        str_val = ''
+        for list_item in str_cols_ix:
+            str_val += str(row[list_item + 1]).lower() + ' '
```

Comment:
if you iterate using "itertuples(index=False)", you do not need to increment list_item